File: packages/node/src/configuration.ts
import { ConfigurationError, MigrationError } from "./errors"
import { isValidSlug, SLUG_RULES_HUMAN } from "./slug-validator"
import {
  SUPPORTED_ENVIRONMENTS,
  detectEnvironment,
  type Environment,
} from "./environment-detector"
import { SDK_VERSION, SDK_LANGUAGE } from "./version"

export interface NurseAndreaConfig {
  orgToken: string
  workspaceSlug: string
  environment: Environment
  host: string
  serviceName: string
  enabled: boolean
  logLevel: "debug" | "info" | "warn" | "error"
  flushIntervalMs: number
  batchSize: number
}

const DEFAULT_HOST = "https://nurseandrea.io"
const LEGACY_FIELDS = ["apiKey", "token", "ingestToken"] as const

let _config: NurseAndreaConfig | null = null
let _bannerPrinted = false

function migrationMessage(field: string): string {
  return (
    `${field} is no longer supported in NurseAndrea SDK 1.0. ` +
    "Migrate to orgToken + workspaceSlug + environment. " +
    "See https://docs.nurseandrea.io/sdk/migration"
  )
}

export function configure(options: Partial<NurseAndreaConfig> & Record<string, unknown>): void {
  for (const legacy of LEGACY_FIELDS) {
    if (options[legacy] !== undefined) {
      throw new MigrationError(migrationMessage(legacy))
    }
  }

  const orgToken = options.orgToken ?? process.env.NURSE_ANDREA_ORG_TOKEN ?? ""
  const workspaceSlug = options.workspaceSlug ?? ""
  const environment = (options.environment ?? detectEnvironment()) as Environment
  const host = options.host ?? process.env.NURSE_ANDREA_HOST ?? DEFAULT_HOST

  if (!orgToken) {
    throw new ConfigurationError("orgToken is required")
  }
  if (!workspaceSlug) {
    throw new ConfigurationError("workspaceSlug is required")
  }
  if (!environment) {
    throw new ConfigurationError("environment is required")
  }
  if (!(SUPPORTED_ENVIRONMENTS as readonly string[]).includes(environment)) {
    throw new ConfigurationError(
      `environment must be one of ${SUPPORTED_ENVIRONMENTS.join(", ")} (got ${JSON.stringify(environment)})`
    )
  }
  if (!isValidSlug(workspaceSlug)) {
    throw new ConfigurationError(
      `workspaceSlug ${JSON.stringify(workspaceSlug)} is invalid. ${SLUG_RULES_HUMAN}`
    )
  }

  _config = {
    orgToken,
    workspaceSlug,
    environment,
    host,
    serviceName:     options.serviceName ?? process.env.NURSE_ANDREA_SERVICE_NAME ?? detectServiceName(),
    enabled:         options.enabled ?? process.env.NODE_ENV !== "test",
    logLevel:        options.logLevel ?? "warn",
    flushIntervalMs: options.flushIntervalMs ?? 5000,
    batchSize:       options.batchSize ?? 100,
  }

  // Defer require to break circular import (client.ts imports from this file).
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { client } = require("./client") as typeof import("./client")
  client.start()

  const { startTraceExporter } = require("./tracing/exporter") as typeof import("./tracing/exporter")
  startTraceExporter()

  if (!_bannerPrinted) {
    _bannerPrinted = true
    process.stdout.write(
      `[NurseAndrea] Shipping to ${_config.host} as ${_config.serviceName} ` +
      `(${SDK_LANGUAGE} sdk v${SDK_VERSION}, workspace=${_config.workspaceSlug}/${_config.environment})\n`
    )

    const stop = () => client.stop()
    process.on("beforeExit", stop)
    process.on("SIGTERM", stop)
  }
}

export function getConfig(): NurseAndreaConfig {
  if (!_config) {
    throw new ConfigurationError(
      "NurseAndrea is not configured. Call configure({ orgToken, workspaceSlug, environment }) at startup."
    )
  }
  return _config
}

export function isEnabled(): boolean {
  return !!_config && _config.enabled && !!_config.orgToken
}

export function ingestUrl(): string {
  return `${getConfig().host.replace(/\/$/, "")}/api/v1/ingest`
}

export function metricsUrl(): string {
  return `${getConfig().host.replace(/\/$/, "")}/api/v1/metrics`
}

export function _resetForTests(): void {
  _config = null
  _bannerPrinted = false
}

function detectServiceName(): string {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const pkg = require(process.cwd() + "/package.json")
    return pkg.name ?? "node-app"
  } catch {
    return "node-app"
  }
}

File: packages/go/nurseandrea/configuration.go
package nurseandrea

import (
	"fmt"
	"os"
	"strings"
	"sync"
)

const defaultHost = "https://nurseandrea.io"

// Config holds all SDK configuration.
//
// In 1.0 the auth contract changed from a single workspace token to:
//
//	OrgToken      — your org's ingest token
//	WorkspaceSlug — which workspace within the org receives ingest
//	Environment   — production | staging | development
//
// Legacy fields (APIKey, Token, IngestToken) remain on this struct only so
// Validate() can detect old usage and return a MigrationError pointing at
// the new contract. Setting any of them is a hard configuration error.
type Config struct {
	OrgToken      string
	WorkspaceSlug string
	Environment   string
	Host          string
	ServiceName   string
	Enabled       *bool
	FlushIntervalMs int
	BatchSize       int

	// Migration trip-wires. Setting any of these returns a MigrationError
	// from Validate / Configure. Do not use.
	APIKey      string `json:"-"`
	Token       string `json:"-"`
	IngestToken string `json:"-"`
}

var (
	globalConfig Config
	configured   bool
	configMu     sync.RWMutex
)

// Configure sets the global SDK configuration. Returns an error if validation
// fails (legacy fields, missing required, unsupported environment, invalid slug).
// Returning an error rather than panicking lets callers integrate this into
// their existing startup error-handling flow.
func Configure(cfg Config) error {
	if err := validateLegacyFields(cfg); err != nil {
		return err
	}

	configMu.Lock()
	defer configMu.Unlock()

	if cfg.OrgToken == "" {
		cfg.OrgToken = os.Getenv("NURSE_ANDREA_ORG_TOKEN")
	}
	if cfg.Host == "" {
		cfg.Host = os.Getenv("NURSE_ANDREA_HOST")
	}
	if cfg.Host == "" {
		cfg.Host = defaultHost
	}
	cfg.Host = strings.TrimRight(cfg.Host, "/")

	if cfg.Environment == "" {
		cfg.Environment = DetectEnvironment()
	}

	if cfg.ServiceName == "" {
		cfg.ServiceName = firstNonEmpty(
			os.Getenv("NURSE_ANDREA_SERVICE_NAME"),
			os.Getenv("RAILWAY_SERVICE_NAME"),
			"go-app",
		)
	}
	if cfg.FlushIntervalMs == 0 {
		cfg.FlushIntervalMs = 5000
	}
	if cfg.BatchSize == 0 {
		cfg.BatchSize = 100
	}
	if cfg.Enabled == nil {
		enabled := true
		cfg.Enabled = &enabled
	}

	if err := validateRequired(cfg); err != nil {
		return err
	}

	globalConfig = cfg
	configured = true

	bannerHost := cfg.Host
	bannerService := cfg.ServiceName
	bannerSlug := cfg.WorkspaceSlug
	bannerEnv := cfg.Environment

	// Trigger lazy client init + tracing exporter, print startup banner.
	go func() {
		GetClient()
		tracingStart()
		if !bannerPrinted {
			bannerPrinted = true
			fmt.Fprintf(os.Stdout,
				"[NurseAndrea] Shipping to %s as %s (%s sdk v%s, workspace=%s/%s)\n",
				bannerHost, bannerService, SDKLanguage, Version, bannerSlug, bannerEnv)
		}
	}()

	return nil
}

func validateLegacyFields(cfg Config) error {
	if cfg.APIKey != "" {
		return newMigrationError("APIKey")
	}
	if cfg.Token != "" {
		return newMigrationError("Token")
	}
	if cfg.IngestToken != "" {
		return newMigrationError("IngestToken")
	}
	return nil
}

func validateRequired(cfg Config) error {
	if cfg.OrgToken == "" {
		return &ConfigurationError{Message: "OrgToken is required"}
	}
	if cfg.WorkspaceSlug == "" {
		return &ConfigurationError{Message: "WorkspaceSlug is required"}
	}
	if cfg.Environment == "" {
		return &ConfigurationError{Message: "Environment is required"}
	}
	if !isSupportedEnvironment(cfg.Environment) {
		return &ConfigurationError{
			Message: fmt.Sprintf("Environment must be one of %v (got %q)",
				SupportedEnvironments, cfg.Environment),
		}
	}
	if !IsValidSlug(cfg.WorkspaceSlug) {
		return &ConfigurationError{
			Message: fmt.Sprintf("WorkspaceSlug %q is invalid. %s",
				cfg.WorkspaceSlug, SlugRulesHuman),
		}
	}
	return nil
}

var bannerPrinted bool

// tracingStartFunc is set by the tracing package to avoid circular imports.
var tracingStartFunc func()

// SetTracingStartFunc is called by the tracing package's init() to register itself.
func SetTracingStartFunc(fn func()) {
	tracingStartFunc = fn
}

func tracingStart() {
	if tracingStartFunc != nil {
		tracingStartFunc()
	}
}

// GetConfig returns the current configuration. Panics if Configure has not
// been called — there are no sensible defaults for OrgToken/WorkspaceSlug.
func GetConfig() Config {
	configMu.RLock()
	defer configMu.RUnlock()
	if !configured {
		panic("nurseandrea: Configure() has not been called. Set OrgToken + WorkspaceSlug + Environment at startup.")
	}
	return globalConfig
}

// IsEnabled returns true if the SDK is active.
func IsEnabled() bool {
	cfg := GetConfig()
	return cfg.Enabled != nil && *cfg.Enabled && cfg.OrgToken != ""
}

// IngestURL returns the full ingest endpoint URL.
func IngestURL() string {
	return GetConfig().Host + "/api/v1/ingest"
}

// MetricsURL returns the full metrics endpoint URL.
func MetricsURL() string {
	return GetConfig().Host + "/api/v1/metrics"
}

// DeployURL returns the full deploy event endpoint URL.
func DeployURL() string {
	return GetConfig().Host + "/api/v1/deploy"
}

// BoolPtr is a helper to create a *bool for Config.Enabled.
func BoolPtr(b bool) *bool {
	return &b
}

// resetForTests is exposed for tests that need to clear configured state.
func resetForTests() {
	configMu.Lock()
	defer configMu.Unlock()
	globalConfig = Config{}
	configured = false
	bannerPrinted = false
	resetEnvWarningForTests()
}
